### DISCUSSION TITLE
More time spent on FlowDataControlQueue

### DISCUSSION THREAD

**Discussion by prav2019** [aio-libs/aiohttp#7329] [Q&A] [answered]
![Screen Shot 2023-06-27 at 1 27 39 PM](https://github.com/aio-libs/aiohttp/assets/51928885/71ffb163-89d4-45ff-b7d7-1ef8eb832b93)

It looks like the async websocket server we are using spending more time in FlowDataControlQueue, what does that indicate? I could reduce it by adding await asyncio.sleep(0) in websocket handlers, but not sure how many I should add ? doesn't make sense in adding everywhere. Could anyone help me on what might be happening ? Also using web.run to start the app

**Comment by prav2019** [comment]
Looks like the aiohttp websocket app degrade over time dropping/ delaying the tasks created from **create_task**  api and eventually need to restart the service once every 2-3 hrs

**Reply by Dreamsorcerer** [reply]
Not sure that has any relation to the graphs shown. What does the code look like? Again, more information is needed if you want help.

**Reply by prav2019** [reply]
I cannot add complete code here, but I will add prototype probably:

`async for msg in websocket_response:
###  creating bg tasks (using asyncio create_task)
await asyncio.sleep(0)
`

**Reply by Dreamsorcerer** [reply]
It's the bit in the middle I'm most interested in. You may not be using `create_task()` correctly. If you want to spin up tasks in the background without waiting on them, then you probably want to use aiojobs instead.

**Comment by prav2019** [comment]
The callgraph to trace the path
![Screen Shot 2023-06-27 at 2 29 31 PM](https://github.com/aio-libs/aiohttp/assets/51928885/dc6d8cf9-c685-4010-b736-a8aae0dc009f)

**Comment by Dreamsorcerer** [comment]
You're going to have to provide some more information on what I'm looking at and what you expect to be happening...

If that graph is showing amount of time the program is at a particular point in the code, then that looks pretty normal. I'm pretty sure the FlowControlDataQueue is where the socket code is handled, so whenever you are waiting on socket data it would be sitting in the code. It is not blocking or causing any performance issues, you application is simply idling at this point.

**Reply by prav2019** [reply]
So that means it is not blocking rather it is waiting there without doing anything. So we need to start the service once every 2-3 hrs. Does that mean we have some leaks ? Is there a condition where tasks created using create_task can get dropped ???? or a socket connection is closed by aiohttp ???? Also should we use web.run or apprunner to start the app ?

**Reply by Dreamsorcerer** [reply]
Again, not enough information for me to know what the problem is. Try using aiojobs as above, maybe that'll help. And, yes, you generally want to use `web.run_app()` as the entry point of your application.

**Reply by prav2019** [reply]
FlowControlDataQueue.read is the function that is where event loop is spending more time on. What might be the possibilities, the time spent increases as hours passes

**Reply by prav2019** [reply]
Want to know if we are accepting more requests than we can process, do we need throttling or something to take in how much the service can process

**Reply by Dreamsorcerer** [reply]
See my first response, I would expect it to spend time there, as it is waiting for the sockets. If the CPU is not too high or something, then it should be fine. Again, if you don't provide any more information about your application, then we can't help.

**Reply by prav2019** [reply]
![Screen Shot 2023-07-06 at 12 32 27 PM](https://github.com/aio-libs/aiohttp/assets/51928885/dee0cb4b-ad35-4a84-9e4b-c05d72d75fba)

**Reply by prav2019** [reply]
those 2 values increases to 70 to 80% i don't see more time spent in other modules

**Reply by prav2019** [reply]
async def retrieve_msg_v1(ws, websocket_response, peer_id):
    ................ other code
    
    ......inside the method body where we are creating bg_task
      if other_peer_id in manager.local_peers:
                            logger.info(f"retrieve_msg_v1() other_peer_id exists: {other_peer_id in manager.local_peers}")
                            other_session_exists = manager.session_exists(other_peer_id, peer_id)
                            logger.info(f"retrieve_msg_v1() session_exists exists: {other_session_exists}")
                            if not other_session_exists:
                                manager.create_session(other_peer_id, peer_id)
                                # how could it already exist?
                            create_bg_task(manager.local_peers[other_peer_id].session_ok(other_peer_id, peer_id))  # no need to wait
                            create_bg_task(ws.session_ok(peer_id, other_peer_id))  # no need to wait
                        else:
                            create_bg_task(REDIS_CLIENT.send_session(peer_id, other_peer_id))


def create_bg_task(coro):
          task = loop.create_task(coro)
          create_bg_task.refs.add(task)  # create a strong ref
          # To prevent keeping references to finished tasks forever,
          # make each task remove its own reference from the set after
          # completion:
          task.add_done_callback(handle_task_result)
          return task

**Reply by prav2019** [reply]
this is the best i can do in sharing on how we are creating tasks in websocket handler **retrieve_msg_v1**

**Reply by prav2019** [reply]
![Screen Shot 2023-07-06 at 12 51 04 PM](https://github.com/aio-libs/aiohttp/assets/51928885/37dba823-bead-495a-9a76-fa78673fc92d)
flowcontroldataqueue read api

**Reply by prav2019** [reply]
the problem gets worse as we get more traffic

**Reply by prav2019** [reply]
We are uploading profile dumps to gcs every 15 min, i see sometimes there is a delay, few seconds delay are fine, but there seems to be a too much delay. So something is not right with tasks

![Screen Shot 2023-07-06 at 1 24 19 PM](https://github.com/aio-libs/aiohttp/assets/51928885/71899b71-2c17-4458-ae79-82ddd6bc9f3f)
![Screen Shot 2023-07-06 at 1 21 10 PM](https://github.com/aio-libs/aiohttp/assets/51928885/203b0a95-5502-47b8-b20c-ea11a1ad28bf)

**Reply by prav2019** [reply]
MONITOR: current tasks: 324 (tasks created using **create_bg_task** the function that's mentioned above) ttlcache size: 290 redis_conn: connections: 8 subs: 12370 max_rss: 81.48MB now_rss: 81.48MB loc_peers (ws conn in the pod): 161

**Reply by prav2019** [reply]
def create_bg_task(coro):
        task = loop.create_task(coro)
        create_bg_task.refs.add(task) # create a strong ref
        # To prevent keeping references to finished tasks forever,
        # make each task remove its own reference from the set after
        # completion:
        task.add_done_callback(handle_task_result)
        return task
FYI again...

**Comment by prav2019** [comment]
[task_starving_test_code.zip](https://github.com/aio-libs/aiohttp/files/11974513/task_starving_test_code.zip)
Here is the test code that i have created to simulate the issue we are having

1. When 200 ws connections are created, the save profiler file task never executes and there will be lot of request handlers tasks in pending state:
![IMG_20230706_211752](https://github.com/aio-libs/aiohttp/assets/51928885/2794f9e4-e09c-48f9-9f8a-e3afaed20d11)
2> The profiler file save task executes after closing client (within few seconds)
3> Adding await asyncio.sleep(0) fixes this problem
![IMG_20230706_205946](https://github.com/aio-libs/aiohttp/assets/51928885/28ae0609-5b1d-4177-b21d-23cb3e5577c8)
(top one with sleeps and bottom one without sleeps)
So are sleeps necessary here when lot of time is spent in request handlers to explicitly give control to execute other tasks.

Since i can reproduce with the code, want to understand what exactly is helping the tasks execution with sleep(s) inside websocket reads (for loop)

**Reply by prav2019** [reply]
@Dreamsorcerer Hope the above helps!!!

**Reply by Dreamsorcerer** [reply]
Still rather difficult to figure anything out given the random snippets of information. At a guess, there are 2 things I can think of.

First, in the case of your sleep() call, this is causing an explicit yield to the event loop and allowing it to run some other code. Maybe the code where you are sleeping is actually taking a fair amount of time and stopping other tasks from running without the sleep(). I'm not familiar with how your tool measures time spent in async functions, so it may also mean that these calls have been started and then get blocked by the busy function before they resume.

Second, create_by_task() is a difficult thing to get right, so I strongly suggest replacing this with aiojobs (there will probably be a new class in asyncio in the next 2 Python releases which can replace it later).

If you want any more help than that, you'll need to create a minimal reproducer that we can run and try out.

**Reply by prav2019** [reply]
https://github.com/aio-libs/aiohttp/issues/7344 I created issue and i have added a sample code where i can replicate the behavior

**Reply by prav2019** [reply]
Looks like more than create task it is linked to the ws reads and await ws.send calls they are noy yielding sometimes, worse during high load. I have explained in the issue with more details , expected behavior and zip file that has code to run and test

**Reply by prav2019** [reply]
[task_starving_test_code.zip](https://github.com/aio-libs/aiohttp/files/11985951/task_starving_test_code.zip)
FYI here is the minimal reproducer code. I have posted logs , expected behavior here: 
https://github.com/aio-libs/aiohttp/issues/7344

**Reply by prav2019** [reply]
@Dreamsorcerer Thank you so much, hope this helps to go through what's happening.

**Comment by bdraco** [comment]
Please give 3.10.2 a try. It should scale a bit better. There is still more to do.

**Comment by tarasko** [comment]
I ended up making and using my own small WebSockets library.
https://github.com/tarasko/picows
Check out the benchmark I made, still a lot of room for improvement in aiohttp.

**Comment by bdraco** [comment, marked as answer]
3.11.x has replaced most of this code. The performance should be significantly improved. Please give it a try.
